test/gtest/common/test_helpers.cc

```diff
@@ -607,4 +607,47 @@ message_stream::~message_stream() {
 
 } // detail
 
+template<typename T>
+void cartesian_product(std::vector<std::vector<T> > &final_output,
```

Comment:
i would expect cartesian to have std::vector<T1> and std::vector<T2> as inputs, and have std::vector< std::pair<T1,T2> > as output

Reply:
what if we would like to generate 3d memtype triples or other?
then it would be impossible

Reply:
so this would generate N-dimensional cartesian?

Reply:
> so this would generate N-dimensional cartesian?

yes, it is possible to generate cartesian product for N vectors with different sizes.
So, the cardinality of the result is `output.size == input[0].size * input[1].size * ... * input[N].size` and `output[i],size == N` where `i = 1..N`